src/typegraph/graph/src/naming/mod.rs

```diff
@@ -0,0 +1,78 @@
+// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
+// SPDX-License-Identifier: MPL-2.0
+
+use std::collections::HashMap;
+
+use crate::conv::ValueType;
+use crate::interlude::*;
+use crate::{Arc, FunctionType, ObjectType, Type, TypeNode as _};
+
+#[derive(Default)]
+pub struct NameRegistry {
+    pub(crate) map: HashMap<Arc<str>, Type>,
+}
+
+impl NameRegistry {
+    pub fn register(&mut self, name: String, ty: Type) -> Result<()> {
+        let name: Arc<str> = name.into();
+        ty.base()
+            .name
+            .set(name.clone())
+            .map_err(|e| eyre!("name cannot be set more more than once: '{e}'"))?;
+        if self.map.insert(name.clone(), ty).is_some() {
+            bail!("failed to register duplicate name: '{}'", name);
+        }
+        Ok(())
+    }
+}
+
+pub trait NamingEngine {
+    fn name_value_types(&mut self, types: &ValueType) -> Result<()>;
+    fn name_function(&mut self, function: &Arc<FunctionType>) -> Result<()>;
+    fn name_ns_object(&mut self, ns_object: &Arc<ObjectType>) -> Result<()>;
+    fn registry(&mut self) -> &mut NameRegistry;
+}
+
```

Comment:
Any reason why you made this generic?

Reply:
To enable other naming conventions. One might want to have different naming conventions for different naming conventions for different metagen clients, introspection, ...
It is not currently used, but (...)...